### ISSUE TITLE
How to pass Vinyl file data to templates?

### ISSUE THREAD

**Issue by janwerkhoven** [carlosl/gulp-nunjucks-render#64]
How does one pass [Vinyl file](https://github.com/gulpjs/vinyl) objects to the templates?

```html
// layout.nunjucks

{{file.path}}
{{file.dirname}}
{{file.relative}}
```
When building navigation menus, internationalising pages and doing SEO, I've often found myself in need for the page to know it's own path. How to do it? I've been breaking my head over this, but can't figure it out.

The closest I've come is to use [Gulp Tap](https://github.com/geejs/gulp-tap) which outputs file data, but I haven't been able to pass this as data to `nunjucksrender`.
```
gulp.task('compileHtml', () => {
  return gulp.src(`src/templates/pages/**/*.+(html|nunjucks)`)
    .pipe(tap(function(file, t) {
      util.log(file.dirname.toString())
    }))
    .pipe(nunjucksRender({
      path: ['src/templates'],
      data: {
        data: myData
      }
    }))
    .pipe(gulp.dest(`dist/`))
});
```
Any ideas?

**Comment by kristijanhusak** [comment]
Where is `myData` defined?
I think you do not need any additional plugins, just get the file information before returning gulp.src, and pass that as data.

**Comment by janwerkhoven** [comment]
So `myData` is defined earlier in the Gulpfile and static for all templates:
```javascript
const myData = {
  projectName: 'This Project Name',
  environment: util.env.env || 'development',
  googleAnalyticsID: 'UA-12345678-XX',
  sitemapRootUrl: 'https://www.thisproject.com'
};
```
This allows templates to embed static data + environment logic:
```html
<h1>{{data.projectName}}</h1>
```
```html
{% if data.environment == 'production' %}
  <meta name="robots" content="index, follow">
{% else %}
  <meta name="robots" content="noindex, nofollow">
{% endif %}
```
Which is cool, but this information **does not change** per file I create. Is there a way to pass to each template it's Vinyl file data?

**Comment by janwerkhoven** [comment]
One example where it would be useful:

In navigation menus the active link should get a class `active`. Could do this by having it check it's own `file.dirname` and match it to the `link.url`.
```javascript
data = {
  links: [
    { label: 'Homepage', url: '/' },
    { label: 'Products', url: '/products' },
    { label: 'Features', url: '/features' },
    { label: 'Contact', url: '/contact' }
  ]
}
```
```html
<nav>
  <ul>
    {% for link in data.links %}
      <li>
        <a href="{{link.url}}">{{link.label}}</a>
      </li>
    {% endfor %}
  </ul>
</nav>
```
Could use something like this??
```
<li class="{{ 'active' if link.url === file.dirname }}">
```

**Comment by mohitsinghs** [comment]
For the similar reasons I derived [gulp-nunjucks-md](https://npm.im/gulp-nunjucks-md) from this plugin. I used front-matter at the top of each file and then attached that to a `page` object at rendering time.
I will define navigation menus in a `data.json` like this.
```json
{
  "links" : [
    { "label": "Homepage", "url": "/" },
    { "label": "Products", "url": "/products" },
    { "label": "Features", "url": "/features" },
    { "label": "Contact", "url": "/contact" }
  ]
}
```
And my page will have a frontmatter like this
```yaml
---
title: "Products"
url: "/products/"
---
```
This frontmatter will be attached to a `page` object for every single page. So now we can easily access `page.title` and `page.url` globally for current page.

 And now I can use them as
```nunjucks
<nav>
  {% for link in links %}
    <li><a class="{% if link.url == page.url %} active{% endif %}" href="{{ link.url }}">{{ link.label }}</a></li>
{% endfor %}
</nav>
```
Long story short, I will have two types of data, **global** _which will be passed as a json file or an object_ and **page specific** _which will be in the frontmatter of the page_ . I hope this will help you.

**Comment by revelt** [comment]
@janwerkhoven well your idea with `gulp.tap` is fine and you're very close. In short, what you need is to `tap` the vinyl variables, dump them into some object somewhere higher in the variable scope, then use `gulp-data` to inject the custom objects into Nunjucks renderer through customising Nunjucks environment variable. Sounds complex but it's not, see my patched up piece below.

I did a test on my code and managed to add `filePath` and `fileRelative` (`fileDirname` was `null` in my case because the file was sitting in the root) into the Nunjucks Environment object. I can use `{{ filePath }}` and others anywhere in the template now.

Here's my patched up version of your code, I could not test it, but snippets should be correct as they've been taken from my working sample:

```js
// declare a global var where we'll dump the `file.*` stuff:
var vinylInfo = {}

// use this to manage the data:
const data = require('gulp-data')

// data importer function
function getDataForFile (file) {
  var finalData = {}
  // start merging in the data:
  finalData = Object.assign(finalData, vinylInfo)
  // you can add more objects into your global object: SCSS vars, settings or whatnot.
  // all these vars will be available within Nunjucks templates using `{{ varName }}`.
  // just Object.assign them here.
  // Feel free to read JSON files from the filesystem and merge them in.
  // Tip. If you need advanced merging where maximum data is retained when there are data clashes,
  // check out https://github.com/codsen/object-merge-advanced from yours truly.
  
  return finalData
}

gulp.task('compileHtml', () => {
  // that's an ES6 arrow function, we create a variable named `manageEnvironment` which is equal to a function.
  var manageEnvironment = environment => {
    // add Nunjucks filters here
    // for example removal of "px" from the variable within Nunjucks:
    // environment.addFilter('removepx', str => str && str.replace(/px/g, '', str))
  
    // Turning off the HTML escaping, so we can put HTML in the content.
    // Saves us from hassle escaping all variables in templates.
    environment.opts.autoescape = false

    // Automatically remove leading whitespace from a block/tag:
    environment.opts.lstripBlocks = true  
  }

  return gulp.src(`src/templates/pages/**/*.+(html|nunjucks)`)
    .pipe(tap(file => {
      vinylInfo.filePath = file.path
      vinylInfo.fileDirname = file.dirname
      vinylInfo.fileRelative = file.relative
    }))
    .pipe(data(getDataForFile))
    .pipe(nunjucksRender({
      path: ['src/templates'],
      manageEnv: manageEnvironment,
      envOptions: {
        throwOnUndefined: false
      }
    }))
    .pipe(gulp.dest(`dist/`))
})
```

Managing the Nunjucks environment is slightly confusing because you have to set some things one way (`manageEnvironment` function) and some thing other way (inline via `envOptions` key). It's all in the documentation(s), but many people mix up the two and get frustrated it does not work. I added some most common settings above for you, just in case you need them in the future.

PS. Check out https://standardjs.com/, you probably don't need semicolons unless your team insists on using them.

**Comment by TheDancingCode** [comment]
This can easily be achieved by using `gulp-data`:
```js
const data = require('gulp-data');

gulp.task('compileHtml', () => {
  return gulp.src(`src/templates/pages/**/*.+(html|nunjucks)`)
    .pipe(data(function(file) {
        return {
          file: {
            path: file.path,
            dirname: file.dirname,
            relative: file.relative
          }
        };
      }))
    .pipe(nunjucksRender({
        path: ['src/templates'],
        data: {
          data: myData
        }
      }))
    .pipe(gulp.dest(`dist/`));
});
```
And now, as desired, you can do this:
```
{{ file.path }}
{{ file.dirname }}
{{ file.relative }}
```
